dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/LoopAnalysis.java

```diff
@@ -0,0 +1,202 @@
+                program.isUnrolled() ? this::findUnrolledLoopsInThread : this::findLoopsInThread;
+
+        program.getThreads().forEach(t -> this.thread2LoopsMap.put(t, loopFindingAlgo.apply(t)));
+    }
+
+    private ImmutableList<LoopInfo> findUnrolledLoopsInThread(Thread thread) {
+        final Map<String, LoopInfo> loopName2InfoMap = new HashMap<>();
+        final List<LoopInfo> loops = new ArrayList<>();
+
+        int loopCounter = 0;
+        for (Event e : thread.getEvents()) {
+            final LoopLabelInfo labelInfo = tryParseLoopLabel(e);
+            if (labelInfo == null) {
+                continue;
+            }
+
+            final String loopName = labelInfo.loopName;
+            if (labelInfo.iterationIndex == 1) {
+                // We start a new loop
+                final LoopInfo loopInfo = new LoopInfo();
+                loopInfo.loopNumber = loopCounter++;
+                loopInfo.loopName = loopName;
+                loopInfo.thread = thread;
+                loopInfo.isUnrolled = true;
+                // NOTE: Inner loops may get copied while unrolling the outer loop.
+                // In that case, the inner loops all may have identical label names
+                // So this put-operation may overwrite an old entry, which is okay!
+                loopName2InfoMap.put(loopName, loopInfo);
+                loops.add(loopInfo);
+            }
+
+            final LoopInfo loopInfo = loopName2InfoMap.get(loopName);
+            if (labelInfo.isBound) {
+                final LoopIterationInfo lastIter = loopInfo.loopIterationInfos.get(loopInfo.loopIterationInfos.size() - 1);
+                // We assume the bound marker to be directly before the loop bounding event.
+                // NOTE: We consider neither the bound marker nor the bounding event to be part of this iteration.
+                lastIter.iterEnd = e.getPredecessor();
+            } else {
+                final LoopIterationInfo newIter = new LoopIterationInfo();
+                newIter.iterNumber = labelInfo.iterationIndex;
+                newIter.iterBegin = e;
+                newIter.containingLoop = loopInfo;
+                loopInfo.loopIterationInfos.add(newIter);
+                Verify.verify(newIter.iterNumber == loopInfo.loopIterationInfos.size());
+
+                if (loopInfo.loopIterationInfos.size() > 1) {
+                    final LoopIterationInfo prevIter = loopInfo.loopIterationInfos.get(loopInfo.loopIterationInfos.size() - 2);
+                    prevIter.iterEnd = e.getPredecessor(); // The previous iteration ends directly before the new starts
+                }
+            }
+        }
+
+        return ImmutableList.copyOf(loops);
+    }
+
+    // A loop-related label has the form
+    // "origLabel.loop/bound OR origLabel.loop/itr_N"
+    private LoopLabelInfo tryParseLoopLabel(Event eventToParse) {
+        if (!(eventToParse instanceof Label && ((Label)eventToParse).getName().contains(".loop"))) {
```

Comment:
`LoopFormVerification` adds the `.loop` suffix, but the suffixes `/bound` and `/itr_N` are generated by `LoopUnrolling`.

You are suggesting to make "loop labels" a sort of first-class citizen by adding methods like `isLoopLabel`, `getIterationNumber`, etc. to `Label`? I kinda dislike cluttering the interface of `Label` with such specific functions that only make sense on a specific subset of labels.
I can move the hardcoded strings all into `LoopUnrolling` though (I think this might be the most canonical place for all loop related stuff)


What I think would be the nicest solution is to have the notion of an `EventAnnotation` which is similar to `CodeAnnotation` but is directly attached to events (each `Event` contains a `List<EventAnnotation>`, similar to tags). Then we could have a `LoopEventAnnotation` that contains all the necessary metadata and we would replace the check by `event.has(LoopEventAnnotation.class)` instead.
But this might better be done in a future PR.

Reply:
This sounds indeed as the nicest solution. But moving the strings to `LoopUnrolling` is good enough for this PR. We can add events annotations in a different PR.